conda/install.py

```diff
@@ -1044,24 +1046,31 @@ def link(prefix, dist, linktype=LINK_HARD, index=None, shortcuts=False):
     has_prefix_files = read_has_prefix(join(info_dir, 'has_prefix'))
     no_link = read_no_link(info_dir)
 
-    with Locked(prefix), Locked(pkgs_dir):
+    # for the lock issue
+    # may run into lock if prefix not exist
+    if not isdir(prefix):
+        os.makedirs(prefix)
+
+    with Locked(prefix):
         for f in files:
             src = join(source_dir, f)
-            dst = join(prefix, f)
-            dst_dir = dirname(dst)
-            if not isdir(dst_dir):
-                os.makedirs(dst_dir)
-            if os.path.exists(dst):
-                log.warn("file already exists: %r" % dst)
-                rm_rf(dst)
-            lt = linktype
-            if f in has_prefix_files or f in no_link or islink(src):
-                lt = LINK_COPY
-            try:
-                _link(src, dst, lt)
-            except OSError as e:
-                sys.exit('failed to link (src=%r, dst=%r, type=%r, error=%r)' %
-                         (src, dst, lt, e))
+            with Locked(src):
+                dst = join(prefix, f)
+                dst_dir = dirname(dst)
+                if not isdir(dst_dir):
+                    os.makedirs(dst_dir)
+                if os.path.exists(dst):
+                    log.warn("file already exists: %r" % dst)
+                    rm_rf(dst)
+                lt = linktype
+                if f in has_prefix_files or f in no_link or islink(src):
+                    lt = LINK_COPY
+                with Locked(dst):
```

Comment:
we already lock prefix, and this directory-lock is generally not working, since it will lock files inside the directory. I think that is why we decided to lock on specific file on yesterday